### ISSUE TITLE
grades.csv sometimes incomplete

### ISSUE THREAD

**Issue by dsanson** [Carnap/Carnap#192]
This has happened to me several times now, but I haven't been able to pin down the circumstances that trigger it.

My usual workflow: 

1.   Export grades.csv from Carnap
2.  Move the file from my downloads folder to my grades folder (overwriting the previous version)
3.  "Refresh" a worksheet in Excel that has been set to treat grades.csv as its data source

Sometimes, but not always, the resulting worksheet will be missing some students. When this has happened, I have then opened grades.csv, and it was also missing those students. So either Excel is doing something weird and clobbering some of the data in grades.csv, or the grades.csv exported from Carnap is incomplete.

Every time this happened, I was able to fix it by going back to Carnap and clicking on the export button again, downloading a new CSV. The new CSV has always been complete. Often (maybe every time?), the new CSV has been available immediately, since I was just going back to the same window in browser that I had used a few minutes earlier. So it seemed like Carnap was using the data that it had already calculated for export the first time I exported.

I'll try to be more careful, and remember to check grades.csv for completeness before refreshing in Excel. That should help determine whether Excel is doing something weird or the problem really is with Carnap's export. And if i happens again, I'll keep copies of both files. And I'll try to see if it is the *same* student records that are left out each time.

But I wanted to post this now, since incomplete grade exports are a bad thing, especially if they go unnoticed!

**Comment by gleachkr** [comment]
Thanks! That'd be really helpful. There's some asynchronous stuff in the grade download that is inherently a little complicated. I can't see where it would be going wrong yet, but it's easy to miss this kind of thing. The fact that you get a working CSV the second time does seem to point in the direction of the download being triggered too quickly.

Here's a link to the relevant logic (it's all regular JavaScript, rather than Haskell):

https://github.com/Carnap/Carnap/blob/8d053d4d9310a2cee196a68a91a5165166ff3d07/Carnap-Server/templates/instructor.julius#L287

If you reproduce, and it turns out that the exported CSV is incomplete (seems probable), it would also be really helpful if you were to pop open the browser inspector and see if there are any JS errors that might point in the direction of the bug.

**Comment by loighic** [comment]
Hi. I was just about to open a new issue when I saw this one. I think that it's the same problem, but I've only been getting it since 11/9 or 11/10.

Anyway, it looks like it is the download beginning too early. When I first select 'export grades', I get a spreadsheet with 73 of the 74 students (and I wait for that spreadsheet while all of the students are "loaded"). The missing student is always the last person alphabetized by last name---although the rows aren't ordered that way in the spreadsheet. Then, if I download again, with no waiting for anything to load, I get all 74 students.

**Comment by gleachkr** [comment]
Thanks @loighic! I touched some of this code in eb27bd22419554cdc7c4f8ff385c2869bfb37a18, on 11/8 (in order to automatically recalculate grades when an extension is applied) and in doing so I may have affected the timing in a way that makes this more reproducible. Does this happen for you every time now?

**Comment by gleachkr** [comment]
OK. I have something that might explain it. I'll deploy it with the next update to the server, hopefully by Monday.

**Comment by loighic** [comment]
Ok, cool. Thanks. And yes, it does happen every time that the full roster of students isn't already loaded when I export grades. 

(Although that said, since the workaround is just to export twice, this isn't really a pressing issue for me.)

**Comment by gleachkr** [comment]
@loighic, I think this should now be fixed. Unfortunately, I think what's fixed was straightforwardly a problem introduced by eb27bd2,so it doesn't seem to help much with tracking down the issue reported by @dsanson.

**Comment by loighic** [comment]
Hmm, strange. I don't know if it will help David---either as a work around or to locate the problem---but I noticed yesterday that if I scroll through my list of students so that each one's total score goes from ‘— ’ to some number, then everyone is loaded and the problem doesn't occur. (Maybe that's obvious, but I figured I'd mention it.)

Anyway, the problem itself seems to be fixed for me, so thanks!
